app/process_proposal.go

```diff
@@ -125,76 +55,78 @@ func (app *App) ProcessProposal(req abci.RequestProcessProposal) abci.ResponsePr
 			// since it has no PFB
 			continue
 		}
 
-		// ensure there is only a single sdk.Msg included in the transaction
-		if len(sdkTx.GetMsgs()) > 1 {
-			logInvalidPropBlock(app.Logger(), req.Header, "invalid PFB found: combined with one or more other sdk.Msg")
-			return abci.ResponseProcessProposal{
-				Result: abci.ResponseProcessProposal_REJECT,
-			}
-		}
-
-		// all PFBs must have a share index, so that we can find their
-		// respective blob.
-		if !isWrapped {
-			logInvalidPropBlock(app.Logger(), req.Header, "Found a MsgPayForBlobs without a share index")
-			return abci.ResponseProcessProposal{
-				Result: abci.ResponseProcessProposal_REJECT,
-			}
-		}
-
-		if err = pfb.ValidateBasic(); err != nil {
-			logInvalidPropBlockError(app.Logger(), req.Header, "invalid MsgPayForBlobs", err)
-			return abci.ResponseProcessProposal{
-				Result: abci.ResponseProcessProposal_REJECT,
-			}
-		}
-
-		for i, shareIndex := range wrappedTx.ShareIndexes {
-			commitment, err := inclusion.GetCommitment(cacher, dah, int(shareIndex), shares.SparseSharesNeeded(pfb.BlobSizes[i]))
-			if err != nil {
-				logInvalidPropBlockError(app.Logger(), req.Header, "commitment not found", err)
-				return abci.ResponseProcessProposal{
-					Result: abci.ResponseProcessProposal_REJECT,
-				}
-			}
-			if !bytes.Equal(pfb.ShareCommitments[i], commitment) {
-				logInvalidPropBlock(app.Logger(), req.Header, "found commitment does not match user's")
-				return abci.ResponseProcessProposal{
-					Result: abci.ResponseProcessProposal_REJECT,
-				}
-			}
+		// validate the blobTx. This is the same validation used in CheckTx ensuring
+		// - there is one PFB
+		// - that each blob has a valid namespace
+		// - that the sizes match
+		// - that the namespaces match between blob and PFB
+		// - that the share commitment is correct
+		if err := blobtypes.ValidateBlobTx(app.txConfig, blobTx); err != nil {
+			logInvalidPropBlockError(app.Logger(), req.Header, fmt.Sprintf("invalid blob tx %d", idx), err)
+			return reject()
 		}
 
+		// validated the PFB signature
 		sdkCtx, err = svHander(sdkCtx, sdkTx, true)
 		if err != nil {
 			logInvalidPropBlockError(app.Logger(), req.Header, "invalid PFB signature", err)
-			return abci.ResponseProcessProposal{
-				Result: abci.ResponseProcessProposal_REJECT,
-			}
+			return reject()
 		}
+
 	}
 
-	return abci.ResponseProcessProposal{
-		Result: abci.ResponseProcessProposal_ACCEPT,
+	// Construct the data square from the block's transactions
+	dataSquare, err := square.Construct(req.BlockData.Txs, appconsts.DefaultMaxSquareSize)
+	if err != nil {
+		logInvalidPropBlockError(app.Logger(), req.Header, "failure to compute data square from transactions:", err)
+		return reject()
+	}
+
+	// Assert that the square size stated by the proposer is correct
+	if dataSquare.Size() != req.BlockData.SquareSize {
+		logInvalidPropBlock(app.Logger(), req.Header, "proposed square size differs from calculated square size")
+		return reject()
```

Comment:
[question]
I think you mentioned it already but I forgot, do we still need the squaresize in the block data?

Reply:
It seems that QGB requires the square size and since we don't persist it anywhere in `celestia-core`, I've left it in the block data itself. In the future when we use a similar data storage scheme as `node`, I think it will become redundant and we can remove it from the block itself. Alternatively, I'm not sure whether the QGB needs the square size (i.e. why is the data root and height not sufficient)?

Reply:
ahh that's right. I believe the square size is added for the qgb's blob inclusion fraud proof mechanism, correct @SweeXordious ?